src/analysis/convex/basic.lean

```diff
@@ -491,6 +491,11 @@ begin
         ht (mem_prod.1 hx).2 (mem_prod.1 hy).2 ha hb hab⟩
 end
 
+lemma convex_pi {ι : Type*} {E : ι → Type*} [Π i, add_comm_monoid (E i)]
+  [Π i, has_scalar 𝕜 (E i)] {s : set ι} {t : Π i, set (E i)} (ht : ∀ i, convex 𝕜 (t i)) :
+  convex 𝕜 (s.pi t) :=
+λ x y hx hy a b ha hb hab i hi, ht i (hx i hi) (hy i hi) ha hb hab
```

Comment:
I have mixed feelings about terms like this.

On the one hand, it's amazing but on the other hand I think a non-formalist could exhibit it as evidence that we're reducing mathematics to symbol shuffling.

Reply:
Arguably the whole `pi` thing is just symbol shuffling anyway.